### ISSUE TITLE
Algorithm support

### ISSUE THREAD

**Issue by CavalloneChen** [KaTeX/KaTeX#1863]
It would be helpful to have algorithm support like this:

```
\begin{algorithm}[H]
 \KwData{this text}
 \KwResult{how to write algorithm with \LaTeX2e }
 initialization\;
 \While{not at end of this document}{
  read current\;
  \eIf{understand}{
   go to next section\;
   current section becomes this one\;
   }{
   go back to the beginning of current section\;
  }
 }
 \caption{How to write algorithms}
\end{algorithm}
```
![image](https://user-images.githubusercontent.com/28266566/53288555-72c2d680-37c4-11e9-8423-687525a65646.png)

**Comment by edemaine** [comment]
I think this would make a neat exercise in how far we can go with the macro system. The example you give looks fairly easy *except* for the inability for custom environments (#977) and I don't know how those vertical lines get implemented (one way involves counters, which we also don't support yet, though that could be easily dealt with using JavaScript functions).

A challenge with algorithm environments in particular is that there are so many different and incompatible choices in LaTeX. https://en.wikibooks.org/wiki/LaTeX/Algorithms lists four, and I've often used clrscode/clrscode3e.  This suggests that separate extensions, each mirroring a specific LaTeX package, might be best.

**Comment by kevinbarabash** [comment]
I assume that this environment will require generating some DOM nodes.  This is the main reason why we haven't exposed even a private version of `defineFunction` or `defineEnvironment`.  Currently, we'd have to expose a lot of other private functions.  It might make sense to extract those functions into a separate npm package and then both katex and the custom environment could use them with katex having to re-export them all.

**Comment by jacanchaplais** [comment]
Sorry to ping an old issue, but I would also appreciate this functionality. Would make my Notion notes a lot nicer. :)

**Comment by mjspeck** [comment]
Algorithm support would be huge for so many people. And another way for KaTeX to assert superiority over MathJax!

**Comment by ShugarSkull** [comment]
This would be very helpful indeed !

**Comment by Plinkett** [comment]
Here just to say that it would indeed be a very nice feature.

**Comment by YudiXie** [comment]
Yes, it would be a really great feature to have! I am also using KaTeX on Notion.

**Comment by vintnes** [comment]
I'm currently hacking this kind of markup together with `gathered` and `align` environments and even a very basic implementation would save me a lot of cruft.

**Comment by samiede** [comment]
+1

**Comment by ebenezergelo** [comment]
+2

**Comment by erensunerr** [comment]
+3

**Comment by COD1995** [comment]
+4

**Comment by PrinOrange** [comment]
Is there anything progress？👀
From 2025
